File: ZMQTest/IFWorker.py
import zmq
from zmq.eventloop.zmqstream import ZMQStream
from tornado.ioloop import IOLoop
from IFCore import IFException, Message, Invocation, IFLoop, IFDefinition
import time
import threading


class IFWorker(object):
    def __init__(self, endpoint, serviceName=None, serviceObject=None, interfaces=[], blocking=True, timeout=None):
        self.__endpoint = endpoint
        socket = zmq.Context().socket(zmq.DEALER)
        self.__stream = ZMQStream(socket, IOLoop.current())
        self.__stream.on_recv(self.__onMessage)
        self.__stream.socket.setsockopt(zmq.LINGER, 0)
        self.__stream.connect(self.__endpoint)
        self.__waitingMap = {}
        self.__waitingMapLock = threading.Lock()
        self.blocking = blocking
        self.timeout = timeout
        self.__isService = serviceName != None
        self.__serviceName = serviceName
        self.__serviceObject = serviceObject
        self.__interfaces = interfaces
        threading.Thread(target=self.__hbLoop, daemon=True).start()
        IFLoop.tryStart()
        if self.__isService:
            self.registerAsService(self.__serviceName, self.__interfaces)

    def __hbLoop(self):
        while True:
            time.sleep(IFDefinition.HEARTBEAT_LIVETIME / 5)
            try:
                if not self.blockingInvoker(timeout=IFDefinition.HEARTBEAT_LIVETIME / 5).heartbeat() \
                        and self.__isService:
                    self.registerAsService(self.__serviceName, self.__interfaces)
            except BaseException as e:
                print('Heartbeat: {}'.format(e))

    def __onMessage(self, msg):
        try:
            # print('msg get: ', msg)
            message = Message(msg)
            # print('msg get: {}'.format(message))
            invocation = message.getInvocation()
            if invocation.isRequest():
                self.__onRequest(message)
            elif invocation.isResponse():
                self.__onResponse(message)
        except BaseException as e:
            import traceback
            exstr = traceback.format_exc()
            print(exstr)

    def __onRequest(self, message):
        # print('on req')
        # print(message)
        sourcePoint = message.fromAddress
        invocation = message.getInvocation()
        try:
            result = invocation.perform(self.__serviceObject)
            responseMessage = Message.newDirectMessage(sourcePoint, Invocation.newResponse(message.messageID, result))
            self.__stream.send_multipart(responseMessage.getContent())
        except BaseException as e:
            errorMsg = Message.newDirectMessage(sourcePoint, Invocation.newError(message.messageID, str(e)))
            self.__stream.send_multipart(errorMsg.getContent())

    def __onResponse(self, message):
        # print('get a response message: ', message)
        invocation = message.getInvocation()
        correspondingID = invocation.getResponseID()
        self.__waitingMapLock.acquire()
        if self.__waitingMap.__contains__(correspondingID):
            (futureEntry, runnable) = self.__waitingMap.pop(correspondingID)
            if invocation.isError():
                futureEntry['error'] = invocation.getError()
            else:
                futureEntry['result'] = invocation.getResult()
            if invocation.hasWarning():
                futureEntry['warning'] = invocation.getWarning()
            # print('ready to run')
            runnable()
        else:
            print('ResponseID not recognized: {}'.format(message))
        self.__waitingMapLock.release()

    def send(self, msg):
        # print('sending: {}'.format(msg))
        self.__stream.send_multipart(msg.getContent())

        id = msg.messageID
        (future, onFinish, resultMap) = InvokeFuture.newFuture()
        self.__waitingMapLock.acquire()
        if self.__waitingMap.__contains__(id):
            raise IFException("MessageID have been used.")
        self.__waitingMap[id] = (resultMap, onFinish)
        self.__waitingMapLock.release()
        return future

    def toMessageInvoker(self, target=None):
        return DynamicRemoteObject(None, toMessage=True, blocking=False, target=target, timeout=None)

    def asynchronousInvoker(self, target=None):
        return DynamicRemoteObject(self, toMessage=False, blocking=False, target=target, timeout=None)

    def blockingInvoker(self, target=None, timeout=None):
        return DynamicRemoteObject(self, toMessage=False, blocking=True, target=target, timeout=timeout)

    @classmethod
    def start(cls):
        IOLoop.instance().start()

    def __getattr__(self, item):
        return InvokeTarget(self, item)


class InvokeTarget:
    def __init__(self, worker, item):
        self.__worker = worker
        self.__name = item

    def __getattr__(self, item):
        item = u'{}'.format(item)
        return self.__defaultInvoker(self.__name).__getattr__(item)

    def __call__(self, *args, **kwargs):
        invoker = self.__defaultInvoker('')
        func = invoker.__getattr__(self.__name)
        return func(*args, **kwargs)

    def __defaultInvoker(self, target):
        if self.__worker.blocking:
            return self.__worker.blockingInvoker(target, self.__worker.timeout)
        else:
            return self.__worker.asynchronousInvoker(target)


class RemoteObject(object):
    def __init__(self, name):
        self.name = name

    def __str__(self):
        return "RemoteObject[{}]".format(self.name)


class DynamicRemoteObject(RemoteObject):
    def __init__(self, worker, toMessage, blocking, target, timeout):
        super(DynamicRemoteObject, self).__init__(target)
        self.__worker = worker
        self.__target = target
        self.__toMessage = toMessage
        self.__blocking = blocking
        self.__timeout = timeout
        self.name = target

    def __getattr__(self, item):
        item = u'{}'.format(item)

        def invoke(*args, **kwargs):
            invocation = Invocation.newRequest(item, args, kwargs)
            if self.__target == '' or self.__target == None:
                message = Message.newBrokerMessage(invocation)
            else:
                message = Message.newServiceMessage(self.__target, invocation)
            if self.__toMessage:
                return message
            elif self.__blocking:
                return self.__worker.send(message).sync(self.__timeout)
            else:
                return self.__worker.send(message)

        return invoke

    def __str__(self):
        return "DynamicRemoteObject[{}]".format(self.name)


class InvokeFuture:
    @classmethod
    def newFuture(cls):
        future = InvokeFuture()
        return (future, future.__onFinish, future.__resultMap)

    def __init__(self):
        self.__done = False
        self.__result = None
        self.__exception = None
        self.__warning = None
        self.__onComplete = None
        self.__metux = threading.Lock()
        self.__resultMap = {}
        self.__awaitSemaphore = threading.Semaphore(0)

    def isDone(self):
        return self.__done

    def isSuccess(self):
        return self.__exception is None

    def result(self):
        return self.__result

    def exception(self):
        return self.__exception

    def warning(self):
        return self.__warning

    def onComplete(self, func):
        self.__metux.acquire()
        self.__onComplete = func
        if self.__done:
            self.__onComplete()
        self.__metux.release()

    def waitFor(self, timeout=None):
        # For Python 3 only.
        # if self.__awaitSemaphore.acquire(True, timeout):
        #     self.__awaitSemaphore.release()
        #     return True
        # else:
        #     return False

        # For Python 2 & 3
        timeStep = 0.1 if timeout is None else timeout / 10
        startTime = time.time()
        while True:
            acq = self.__awaitSemaphore.acquire(False)
            if acq:
                return acq
            else:
                passedTime = time.time() - startTime
                if (timeout is not None) and (passedTime >= timeout):
                    return False
                time.sleep(timeStep)

    def sync(self, timeout=None):
        if self.waitFor(timeout):
            if self.isSuccess():
                return self.__result
            elif isinstance(self.__exception, BaseException):
                raise self.__exception
            else:
                raise IFException('Error state in InvokeFuture.')
        else:
            raise IFException('Time out!')

    def __onFinish(self):
        self.__done = True
        if self.__resultMap.__contains__('result'):
            self.__result = self.__resultMap['result']
        if self.__resultMap.__contains__('warning'):
            self.__warning = self.__resultMap['warning']
        if self.__resultMap.__contains__('error'):
            self.__exception = IFException(self.__resultMap['error'])
        if self.__onComplete is not None:
            self.__onComplete()
        self.__awaitSemaphore.release()


if __name__ == '__main__':
    import threading
    import IFCore

    worker = IFWorker("tcp://127.0.0.1:25034", serviceObject=None, interfaces=['TestInterface 1', 'TestInterface 2'],
                      timeout=1)
    print(worker.listServiceNames())

File: InteractionFreeScala/src/main/scala/com/interactionfree/IFWorker.scala
package com.interactionfree

import java.lang.Thread.UncaughtExceptionHandler
import java.nio.charset.Charset
import java.util.concurrent.{Executors, LinkedBlockingQueue, ThreadFactory}
import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger, AtomicLong}
import scala.language.dynamics
import org.zeromq.{SocketType, ZContext, ZMsg}
import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer
import scala.concurrent.duration._
import scala.concurrent.{Await, ExecutionContext, Future}
import scala.language.postfixOps

object IFDefinition {
  val PROTOCOL = "IF1"
  val DISTRIBUTING_MODE_BROKER = "Broker"
  val DISTRIBUTING_MODE_DIRECT = "Direct"
  val DISTRIBUTING_MODE_SERVICE = "Service"
  val HEARTBEAT_LIVETIME = 10 * 1000
}

object Message {
  private val MessageIDs = new AtomicLong(0)

  def parse(content: Array[Array[Byte]]) = {
    val protocol = new String(content(1), "UTF-8")
    if (!IFDefinition.PROTOCOL.contains(protocol)) throw new IFException(s"Prorocol not available: ${protocol}")
    val messageID = content(2)
    val isOutgoingMessage = content.size == 7
    val distributingMode = if (isOutgoingMessage) new String(content(3), "UTF-8") else "Received"
    val remoteAddress = if (isOutgoingMessage) content(4) else content(4)
    val serialization = new String(content(if (isOutgoingMessage) 5 else 4), "UTF-8")
    val invocation = Invocation.deserialize(content(if (isOutgoingMessage) 6 else 5), serialization)
    new Message(messageID, isOutgoingMessage, distributingMode, remoteAddress, serialization, invocation)
  }

  private def nextMesssageID = MessageIDs.getAndIncrement().toString.getBytes("UTF-8")

  def newBrokerMessage(invocation: Invocation, serialization: String = "Msgpack") = new Message(nextMesssageID, true, IFDefinition.DISTRIBUTING_MODE_BROKER, Array[Byte](), serialization, invocation)

  def newServiceMessage(serviceName: String, invocation: Invocation, serialization: String = "Msgpack") = new Message(nextMesssageID, true, IFDefinition.DISTRIBUTING_MODE_SERVICE, serviceName.getBytes("UTF-8"), serialization, invocation)

  def newDirectMessage(address: Array[Byte], invocation: Invocation, serialization: String = "Msgpack") = new Message(nextMesssageID, true, IFDefinition.DISTRIBUTING_MODE_DIRECT, address, serialization, invocation)

  def newFromBrokerMessage(messageID: Array[Byte], fromAddress: Array[Byte], invocation: Invocation, serialization: String = "Msgpack") = new Message(messageID, false, IFDefinition.DISTRIBUTING_MODE_DIRECT, fromAddress, serialization, invocation)
}

class Message private(val messageID: Array[Byte], val isOutgoingMessage: Boolean, val distributingMode: String, val remoteAddress: Array[Byte], val serialization: String, val invocation: Invocation) {
  val protocol = "IF1"
  val isBrokerMessage = isOutgoingMessage && distributingMode == IFDefinition.DISTRIBUTING_MODE_BROKER
  val isServiceMessage = isOutgoingMessage && distributingMode == IFDefinition.DISTRIBUTING_MODE_SERVICE
  val isDirectMessage = isOutgoingMessage && distributingMode == IFDefinition.DISTRIBUTING_MODE_DIRECT
  private val creationTime = System.currentTimeMillis()

  def expired(lifetime: Long) = System.currentTimeMillis() - creationTime > lifetime

  def messageIDasLong = new String(messageID, "UTF-8").toInt

  override def toString: String = s"[$distributingMode - $remoteAddress] $invocation"
}

object IFWorker {
  def apply(endpoint: String, serviceName: String = "", serviceObject: Any = None, interfaces: List[String] = Nil, timeout: Duration = 10 seconds) = new BlockingIFWorker(endpoint, serviceName, serviceObject, interfaces, timeout)

  def async(endpoint: String, serviceName: String = "", serviceObject: Any = None, interfaces: List[String] = Nil) = new AsyncIFWorker(endpoint, serviceName, serviceObject, interfaces)
}

class IFWorker(endpoint: String, serviceName: String = "", serviceObject: Any = None, interfaces: List[String] = Nil, val timeout: Duration = 10 seconds) {
  val isService = serviceName != ""
  val context = new ZContext()
  val socket = context.createSocket(SocketType.DEALER)
  socket.connect(endpoint)

  private val receivingExecutor = Executors.newSingleThreadExecutor()
  private val hbExecutor = Executors.newSingleThreadExecutor()
  private val localInvokingExecutor = Executors.newSingleThreadExecutor()
  private val sendingQueue = new LinkedBlockingQueue[Message]()
  private val messageExpiration = new AtomicLong(2000)
  private val heartbeatInterval = IFDefinition.HEARTBEAT_LIVETIME / 5
  private val closed = new AtomicBoolean(false)

  def send(msg: Message) = future(msg)

  receivingExecutor.submit(new Runnable {
    private val charset = Charset.forName("UTF-8")

    override def run(): Unit = {
      while (!closed.get) {
        try {
          val zmsg = ZMsg.recvMsg(socket)
          if (zmsg.size() != 6) throw new IFException("Invalid Message from Broker.")
          val it = zmsg.iterator()
          it.next()
          assert(it.next().getString(charset) == IFDefinition.PROTOCOL)
          val messageID = it.next().getData
          val fromAddress = it.next().getData
          val serialization = it.next().getString(charset)
          val invocation = Invocation.deserialize(it.next().getData, serialization)
          val msg = Message.newFromBrokerMessage(messageID, fromAddress, invocation, serialization)
          if (invocation.isRequest) onRequest(msg) else onResponse(msg)
        } catch {
          case e: Throwable => if (!closed.get) println(e)
        }
      }
    }
  })
  hbExecutor.submit(new Runnable {
    override def run(): Unit = {
      while (!closed.get) {
        try {
          val hb = blockingInvoker("", timeout = (IFDefinition.HEARTBEAT_LIVETIME / 5) milliseconds).heartbeat().asInstanceOf[Boolean]
          if (!hb && isService) blockingInvoker().registerAsService(serviceName, interfaces)
        } catch {
          case e: Throwable => if (!closed.get) println(s"Heartbeat: ${e.getMessage}")
        }
        Thread.sleep(IFDefinition.HEARTBEAT_LIVETIME / 5)
      }
    }
  })

  def toMessageInvoker(target: String = "") = new MessageRemoteObject(this, target)

  def asynchronousInvoker(target: String = "") = new AsynchronousRemoteObject(this, target)

  def blockingInvoker(target: String = "", timeout: Duration = 10 second) = new BlockingRemoteObject(this, target, timeout = timeout)

  def close() = {
    closed set true
    if (isService) asynchronousInvoker().unregister()
    socket.close()
    context.close()
    receivingExecutor.shutdown()
    hbExecutor.shutdown()
    localInvokingExecutor.shutdown()
  }

  private class FutureEntry(var result: Option[Any] = None, var cause: Option[Throwable] = None)

  private object SingleThreadExecutionContext extends ExecutionContext with UncaughtExceptionHandler {
    private lazy val SingleThreadPool = Executors.newSingleThreadExecutor(new ThreadFactory() {
      private lazy val ThreadCount = new AtomicInteger(0)

      def newThread(runnable: Runnable): Thread = {
        val t = new Thread(runnable)
        t.setDaemon(true)
        t.setName(s"SingleThreadExecutionContextThread-${ThreadCount.getAndIncrement}")
        t.setUncaughtExceptionHandler(SingleThreadExecutionContext.this)
        t.setPriority(Thread.NORM_PRIORITY)
        t
      }
    })

    def execute(runnable: Runnable): Unit = {
      SingleThreadPool.submit(runnable).get
    }

    def reportFailure(cause: Throwable): Unit = {
    }

    def uncaughtException(thread: Thread, cause: Throwable): Unit = {
    }
  }

  private val waitingMap: mutable.HashMap[Long, (FutureEntry, Runnable)] = new mutable.HashMap

  private def future(msg: Message) = {
    val id = msg.messageIDasLong
    val futureEntry = new FutureEntry
    Future[Any] {
      if (futureEntry.cause.isDefined) throw futureEntry.cause.get
      if (futureEntry.result.isDefined) futureEntry.result.get
      else throw new IFException("Error state: FutureEntry not defined.")
    }(new ExecutionContext {
      def execute(runnable: Runnable): Unit = {
        SingleThreadExecutionContext.execute(() => {
          waitingMap.synchronized {
            if (waitingMap.contains(id)) throw new IFException("MessageID have been used.")
            waitingMap.put(id, (futureEntry, runnable))
          }
          val zmsg = new ZMsg()
          zmsg.addLast("").addLast(msg.protocol).addLast(msg.messageID).addLast(msg.distributingMode).addLast(msg.remoteAddress).addLast(msg.serialization).add(msg.invocation.serialize(msg.serialization))
          val suc = zmsg.send(socket)
        })
      }

      def reportFailure(cause: Throwable): Unit = {
        if (!closed.get) println(s"nn:$cause")
      }
    })
  }

  private def onRequest(msg: Message) = {
    val sourcePoint = msg.remoteAddress
    val invocation = msg.invocation
    localInvokingExecutor.submit(new Runnable {
      override def run(): Unit = {
        try {
          val result = invocation.perform(serviceObject)
          send(Message.newDirectMessage(sourcePoint, Invocation.newResponse(msg.messageID, result)))
        } catch {
          case e: Throwable => send(Message.newDirectMessage(sourcePoint, Invocation.newError(msg.messageID, e.getMessage)))
        }
      }
    })
  }

  private def onResponse(msg: Message) = {
    val invocation = msg.invocation
    val id = new String(invocation.getResponseID, "UTF-8").toLong
    waitingMap.remove(id) match {
      case Some((futureEntry, runnable)) => {
        if (invocation.isError) futureEntry.cause = Some(new IFException(invocation.getError))
        else if (invocation.isResponse) futureEntry.result = Some(invocation.getResult)
        runnable.run()
      }
      case None => if (!closed.get) println(s"MessageID not found: $id")
    }
  }
}

class BlockingIFWorker(endpoint: String, serviceName: String = "", serviceObject: Any = None, interfaces: List[String] = Nil, timeout: Duration = 10 seconds) extends IFWorker(endpoint, serviceName, serviceObject, interfaces, timeout) with Dynamic {
  def selectDynamic(targetName: String): BlockingRemoteObject = blockingInvoker(targetName)

  def applyDynamic(functionName: String)(args: Any*): Any = blockingInvoker("", timeout).applyDynamic(functionName)(args: _*)

  def applyDynamicNamed(functionName: String)(args: (String, Any)*): Any = blockingInvoker("", timeout).applyDynamicNamed(functionName)(args: _*)
}

class AsyncIFWorker(endpoint: String, serviceName: String = "", serviceObject: Any = None, interfaces: List[String] = Nil) extends IFWorker(endpoint, serviceName, serviceObject, interfaces, Duration.Zero) with Dynamic {
  def selectDynamic(targetName: String): AsynchronousRemoteObject = asynchronousInvoker(targetName)

  def applyDynamic(functionName: String)(args: Any*): Future[Any] = asynchronousInvoker("").applyDynamic(functionName)(args: _*)

  def applyDynamicNamed(functionName: String)(args: (String, Any)*): Future[Any] = asynchronousInvoker("").applyDynamicNamed(functionName)(args: _*)
}

class MessageRemoteObject(worker: IFWorker, targetName: String = "") extends Dynamic {
  def selectDynamic(functionName: String): Message = new InvokeItem(worker, targetName, functionName)().toMessage

  def applyDynamic(functionName: String)(args: Any*): Message = new InvokeItem(worker, targetName, functionName)(args.map(m => ("", m)): _*).toMessage

  def applyDynamicNamed(functionName: String)(args: (String, Any)*): Message = new InvokeItem(worker, targetName, functionName)(args: _*).toMessage
}

class BlockingRemoteObject(worker: IFWorker, targetName: String = "", timeout: Duration = 10 second) extends Dynamic {
  def selectDynamic(functionName: String): Any = new InvokeItem(worker, targetName, functionName)().requestMessage(timeout)

  def applyDynamic(functionName: String)(args: Any*): Any = new InvokeItem(worker, targetName, functionName)(args.map(m => ("", m)): _*).requestMessage(timeout)

  def applyDynamicNamed(functionName: String)(args: (String, Any)*): Any = new InvokeItem(worker, targetName, functionName)(args: _*).requestMessage(timeout)
}

class AsynchronousRemoteObject(worker: IFWorker, targetName: String = "") extends Dynamic {
  def selectDynamic(functionName: String): Future[Any] = new InvokeItem(worker, targetName, functionName)().sendMessage

  def applyDynamic(functionName: String)(args: Any*): Future[Any] = new InvokeItem(worker, targetName, functionName)(args.map(m => ("", m)): _*).sendMessage

  def applyDynamicNamed(functionName: String)(args: (String, Any)*): Future[Any] = new InvokeItem(worker, targetName, functionName)(args: _*).sendMessage
}

private class InvokeItem(worker: IFWorker, target: String, functionName: String)(args: (String, Any)*) {
  val argsList: ArrayBuffer[Any] = new ArrayBuffer
  val namedArgsMap: mutable.HashMap[String, Any] = new mutable.HashMap
  args.foreach(m => m match {
    case (name, value) if name == null || name.isEmpty => argsList += value
    case (name, value) => namedArgsMap.put(name, value)
  })
  val invocation = Invocation.newRequest(functionName, argsList.toList, namedArgsMap.toMap)

  def sendMessage = worker.send(toMessage)

  def requestMessage(timeout: Duration) = Await.result[Any](worker.send(toMessage), timeout)

  def toMessage = if (target == "") Message.newBrokerMessage(invocation) else Message.newServiceMessage(target, invocation)
}

object IFWorkerApp extends App {
  val worker = IFWorker("tcp://localhost:224", "IFWorkerAppTest")
  Thread.sleep(5000)
  worker.close()
}